aten/src/ATen/native/eigen/SparseBlas.cpp

```diff
@@ -0,0 +1,263 @@
+        typedef Eigen::SparseMatrix<scalar_t, Eigen::ColMajor, index_t>
+            EigenCscMatrix;
+        typedef Eigen::SparseMatrix<scalar_t, Eigen::RowMajor, index_t>
+            EigenCsrMatrix;
+
+        at::Tensor mat1_mat2 = at::empty(result.sizes(), result.options());
+
+        if (mat1.layout() == kSparseCsr) {
+          const Eigen::Map<EigenCsrMatrix> mat1_eigen =
+              to_EigenCsr<scalar_t, index_t>(mat1);
+          if (mat2.layout() == kSparseCsr) {
+            const Eigen::Map<EigenCsrMatrix> mat2_eigen =
+                to_EigenCsr<scalar_t, index_t>(mat2);
+            const EigenCsrMatrix mat1_mat2_eigen = (mat1_eigen * mat2_eigen);
+            EigenCsr_to_tensor<scalar_t, index_t>(mat1_mat2, mat1_mat2_eigen);
+          } else if (mat2.layout() == kSparseCsc) {
+            const Eigen::Map<EigenCscMatrix> mat2_eigen =
+                to_EigenCsc<scalar_t, index_t>(mat2);
+            const EigenCsrMatrix mat1_mat2_eigen = (mat1_eigen * mat2_eigen);
+            EigenCsr_to_tensor<scalar_t, index_t>(mat1_mat2, mat1_mat2_eigen);
+          }
+        } else if (mat1.layout() == kSparseCsc) {
+          const Eigen::Map<EigenCscMatrix> mat1_eigen =
+              to_EigenCsc<scalar_t, index_t>(mat1);
+          if (mat2.layout() == kSparseCsc) {
+            const Eigen::Map<EigenCscMatrix> mat2_eigen =
+                to_EigenCsc<scalar_t, index_t>(mat2);
+            const EigenCsrMatrix mat1_mat2_eigen = (mat1_eigen * mat2_eigen);
+            EigenCsr_to_tensor<scalar_t, index_t>(mat1_mat2, mat1_mat2_eigen);
+          } else if (mat2.layout() == kSparseCsr) {
+            const Eigen::Map<EigenCsrMatrix> mat2_eigen =
+                to_EigenCsr<scalar_t, index_t>(mat2);
+            const EigenCsrMatrix mat1_mat2_eigen = (mat1_eigen * mat2_eigen);
+            EigenCsr_to_tensor<scalar_t, index_t>(mat1_mat2, mat1_mat2_eigen);
+          }
+        }
+
+        result.add_(mat1_mat2, alpha.to<scalar_t>());
+      });
+}
+
+} // anonymus namespace
+
+void addmm_out_sparse(
+    const at::Tensor& mat1,
+    const at::Tensor& mat2,
+    const at::Tensor& result,
+    const at::Scalar& alpha,
+    const at::Scalar& beta) {
+  TORCH_CHECK(
+      result.layout() == kSparseCsr && mat1.layout() != kStrided && mat2.layout() != kStrided,
+      "eigen::addmm_out_sparse: computation on CPU is not implemented for ",
+      result.layout(),
+      " + ",
+      mat1.layout(),
+      " @ ",
+      mat2.layout());
+
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(
+    result.scalar_type(), "addmm_out_sparse_csr_eigen", [&] {
+      addmm_out_sparse_csr_eigen<scalar_t>(mat1, mat2, result, alpha, beta);
+  });
+}
+
+void add_out_sparse(
+    const at::Tensor& mat1,
+    const at::Tensor& mat2,
+    const at::Scalar& alpha,
+    const at::Tensor& result) {
+  TORCH_CHECK(
+      result.layout() == kSparseCsr && mat1.layout() == kSparseCsr && mat2.layout() == kSparseCsr,
```

Comment:
I think Eigen only requires all operands to have the same compression axis for addition, so this could support CSC + CSC also right?

Reply:
Yes, you are right. Eigen supports CSC + CSC. The limitation is from pytorch side. All my attempts resizing CSC tensor ended up being a segfault. Is there a way to resize CSC tensor without transposing it to CSR so I can resize it like below?
```C++
    static_cast<SparseCscTensorImpl*>(input.unsafeGetTensorImpl())
        ->set_member_tensors(
            input.row_indices().resize_({nnz}),
            input.ccol_indices(),
            input.values().resize_({nnz}),
            input.sizes());
```

Reply:
Yes, it should look like this:

```c++
    static_cast<SparseCsrTensorImpl*>(input.unsafeGetTensorImpl())
        ->set_member_tensors(
            input.ccol_indices(),
            input.row_indices().resize_({nnz}),
            input.values().resize_({nnz}),
            input.sizes());
```

Note that the `impl` type is `SparseCsrTensorImpl` as that impl backs all sparse-compressed layouts (its name is misleading, but it pre-dates having other sparse-compressed layouts). Also I have permuted the first two arguments as the compressed indices should be first and the plain indices second.

Reply:
Thank you! Now the CSC + CSC addition is also supported.